### PR TITLE
rustls-libssl: init at 0.2.1

### LINKED ISSUE #310173: Package request: rustls-libssl
**Project description**

rustls-libssl is a partial reimplementation of the OpenSSL 3 libssl ABI. It is written in rust and uses rustls. It does not replace libcrypto: this is still required.

https://www.memorysafety.org/blog/rustls-nginx-compatibility-layer/

**Metadata**

* homepage URL: https://github.com/rustls/rustls-openssl-compat/tree/main/rustls-libssl
* source URL: https://github.com/rustls/rustls-openssl-compat/tree/main/rustls-libssl
* license: asl20
* platforms: unix, linux, darwin

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by stephank**
Fixes #310173

I've not been able to make this work on Darwin yet. The lib builds, but the symbols are mangled incorrectly for Nginx to find them it seems. Upstream build infra doesn't really seem to support this any way.

I assume this just works on `x86_64-linux`, but haven't tested.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/ru/rustls-libssl/package.nix [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+      -e "s|${openssl.out}|$out|g" \
+      -e "s|${openssl.dev}|$dev|g" \
+      $dev/lib/pkgconfig/*.pc
+  '';
+
+  meta = {
+    description = "partial reimplementation of the OpenSSL 3 libssl ABI using rustls";
+    homepage = "https://github.com/rustls/rustls-openssl-compat";
+    license = lib.licenses.asl20;
+    maintainers = with lib.maintainers; [ stephank ];
```
Comment (cpu) [inline pkgs/by-name/ru/rustls-libssl/package.nix:78]: if you'd like to add a 2nd maintainer I'd be happy to volunteer. My GH nick is already in `lib.maintainers`.

**Review by cpu — APPROVED** [review]
Thanks for putting this together!

> I've not been able to make this work on Darwin yet. The lib builds, but the symbols are mangled incorrectly for Nginx to find them it seems. Upstream build infra doesn't really seem to support this any way.

Indeed - to date we've been focused on Linux exclusively in the upstream project. I think it's probably best to match that here until there's been time to invest in MacOS support upstream.

> I assume this just works on x86_64-linux, but haven't tested.

FWIW I tested both the derivation and the nixos test with success on my x86_64 Linux NixOS system.

File: pkgs/by-name/ru/rustls-libssl/package.nix (hunk 1 above)
Comment (cpu) [reply pkgs/by-name/ru/rustls-libssl/package.nix]: That's perfect, thanks!

**Review by cpu — APPROVED** [review]

File: nixos/tests/all-tests.nix [hunk 2]
```diff
@@ -892,6 +892,7 @@ in {
   rsyncd = handleTest ./rsyncd.nix {};
   rsyslogd = handleTest ./rsyslogd.nix {};
   rtorrent = handleTest ./rtorrent.nix {};
+  rustls-libssl = handleTest ./rustls-libssl.nix {};
```
Comment (FliegendeWurst) [inline nixos/tests/all-tests.nix:895]: You could add this to passthru.tests of the package, then ofborg will build it automatically :)

File: pkgs/by-name/ru/rustls-libssl/package.nix [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+  nativeBuildInputs = [
+    pkg-config # for openssl-sys
+    llvmPackages.lld # build.rs specifies LLD as linker
+  ];
+  buildInputs = [
+    openssl
+  ];
+
+  doCheck = false; # TODO: can't find libcrypto
```
Comment (FliegendeWurst) [inline pkgs/by-name/ru/rustls-libssl/package.nix:40]:
```suggestion
  preCheck = ''
    # tests dlopen libcrypto.so.3
    export LD_LIBRARY_PATH=${lib.makeLibraryPath [ openssl ]}
  '';
```
If the tests are indeed supposed to open openssl's libcrypto.

File: pkgs/by-name/ru/rustls-libssl/package.nix [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    cp ${openssl.dev}/lib/pkgconfig/*.pc $dev/lib/pkgconfig/
+    sed -i \
+      -e "s|${openssl.out}|$out|g" \
+      -e "s|${openssl.dev}|$dev|g" \
+      $dev/lib/pkgconfig/*.pc
+  '';
+
+  meta = {
+    description = "Partial reimplementation of the OpenSSL 3 libssl ABI using rustls";
+    homepage = "https://github.com/rustls/rustls-openssl-compat";
```
Comment (FliegendeWurst) [inline pkgs/by-name/ru/rustls-libssl/package.nix:76]:
```suggestion
    homepage = "https://github.com/rustls/rustls-openssl-compat";
    changelog = "https://github.com/rustls/rustls-openssl-compat/releases";
```
(or link to version if you prefer)

File: pkgs/by-name/ru/rustls-libssl/package.nix (hunk 3 above)
Comment (stephank) [reply pkgs/by-name/ru/rustls-libssl/package.nix]: Oh, great suggestion, that did the trick!

File: nixos/tests/all-tests.nix (hunk 2 above)
Comment (stephank) [reply nixos/tests/all-tests.nix:912]: Good point. Ofborg ran the test successfully. 👍 
https://logs.ofborg.org/?key=nixos/nixpkgs.363932&attempt_id=c521ecbe-c990-45d8-9984-390fc7331ba3

**Comment by stephank** [comment]
Thanks for the review, I added all the suggestions.

**Comment by FliegendeWurst** [comment]
So I'm wondering, can this package simply be used like `somePackage.override { openssl = rustls-libssl; }` or is it more involved?

**Comment by stephank** [comment]
> So I'm wondering, can this package simply be used like `somePackage.override { openssl = rustls-libssl; }` or is it more involved?
> 
> Edit: just tried it for curl and nginx. curl complains about missing `SSL_get0_group_name` but nginx builds.

Yes, that's the intended usage. Not sure why curl doesn't work, because upstream does list it as tested on Ubuntu. But it definitely doesn't cover the full API surface of OpenSSL right now.

**Comment by FliegendeWurst** [comment]
> Not sure why curl doesn't work, because upstream does list it as tested on Ubuntu.

Probably nixpkgs has a newer version.

**Pull request merged**